rcl/include/rcl/client.h

```diff
@@ -0,0 +1,300 @@
+ *    rcl_client_options_t client_ops = rcl_client_get_default_options();
+ *    ret = rcl_client_init(&client, &node, ts, "chatter", &client_ops);
+ *    // ... error handling, and on shutdown do finalization:
+ *    ret = rcl_client_fini(&client, &node);
+ *    // ... error handling for rcl_client_fini()
+ *    ret = rcl_node_fini(&node);
+ *    // ... error handling for rcl_deinitialize_node()
+ *
+ * This function is not thread-safe.
+ *
+ * \param[inout] client preallocated client structure
+ * \param[in] node valid rcl node handle
+ * \param[in] type_support type support object for the topic's type
+ * \param[in] service_name the name of the service to request
+ * \param[in] options client options, including quality of service settings
+ * \return RCL_RET_OK if the client was initialized successfully, or
+ *         RCL_RET_NODE_INVALID if the node is invalid, or
+ *         RCL_RET_ALREADY_INIT if the client is already initialized, or
+ *         RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ *         RCL_RET_BAD_ALLOC if allocating memory fails, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_client_init(
+  rcl_client_t * client,
+  const rcl_node_t * node,
+  const rosidl_service_type_support_t * type_support,
+  const char * service_name,
+  const rcl_client_options_t * options);
+
+/// Finalize a rcl_client_t.
+/* After calling, the node will no longer be advertising that it is publishing
+ * on this topic (assuming this is the only client on this topic).
+ *
+ * After calling, calls to rcl_publish will fail when using this client.
+ * However, the given node handle is still valid.
+ *
+ * This function is not thread-safe.
+ *
+ * \param[inout] client handle to the client to be finalized
+ * \param[in] node handle to the node used to create the client
+ * \return RCL_RET_OK if client was finalized successfully, or
+ *         RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_client_fini(rcl_client_t * client, rcl_node_t * node);
+
+/// Return the default client options in a rcl_client_options_t.
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_client_options_t
+rcl_client_get_default_options(void);
+
+/// Send a ROS request using a client.
+/* It is the job of the caller to ensure that the type of the ros_request
+ * parameter and the type associate with the client (via the type support)
+ * match.
+ * Passing a different type to send_request produces undefined behavior and cannot
+ * be checked by this function and therefore no deliberate error will occur.
+ *
+ * send_request is an non-blocking call. 
+ *
+ * The ROS request message given by the ros_request void pointer is always owned by the
+ * calling code, but should remain constant during publish.
+ *
+ * TODO: Is this true for send_request?
```

Comment:
I think it should be, but of course that depends on the implementation. If there's a compelling reason for it not to be true then we can change this.